### PR TITLE
[WIP] Adam optimizer

### PR DESCRIPTION
**Pull request by kashif**

### CODE REVIEW

File: neon/backends/gpu.py [hunk 1]
```diff
@@ -1154,6 +1154,20 @@ def ada_update(self, ps_item, us_item, gs_item, ds_item, ls_item, ss_item,
...
+    def adam_update(self, ps_item, us_item, ms_item, vs_item, ls_item, ss_item,
+                   beta_1, beta_2, epsilon, epoch):
+
+        ms_item[:] = ms_item * beta_1 + (1.0 - beta_1) * us_item
+
+        vs_item[:] = vs_item * beta_2 + (1.0 - beta_2) * us_item * us_item
+
+        t = epoch + 1
+
+        ls_item[:] = ls_item * self.ng.sqrt(1.0 - beta_2**t) / (1.0 - beta_1**t)
```
Comment (nervanasys) [inline neon/backends/gpu.py]: This could be computed outside this function and doesn't need to be an array (just a scalar)

**Comment by nervanasys** [comment]
Would be nice to have a working MNIST example using Adam.

Btw, please be aware that there will be upcoming changes due to https://github.com/NervanaSystems/neon/issues/62
Nevertheless, we can use your implementation as a template to get Adam working in the refactored neon.

**Comment by kashif** [comment]
thanks @nervanasys I'll have a look at this...

**Comment by kashif** [comment]
closing this since it is in the new refactored version.

**Pull request closed without merging**